088-rebalance-progress-status.md

```diff
@@ -0,0 +1,402 @@
+
+## Current Situation
+
+At this time, Strimzi users are able to execute partition rebalances via `KafkaRebalance` custom resources but can only monitor the progression of those partition rebalances in two ways:
+
+- Manually querying the Cruise Control REST API endpoint directly (which requires creating a custom Cruise Control REST API user or disabling HTTP Basic authentication).
+- Inspecting the logs of the Cruise Control instance. 
+
+Unfortunately, neither of these methods are particularly user-friendly.
+
+## Motivation
+
+Information surrounding the progress of an executing partition rebalance is useful for planning future cluster operations. 
+Knowing things like how much time an ongoing partition rebalance has left to take and how much data an ongoing partition rebalance has left to transfer helps users understand the cost of an ongoing partition rebalance.
+This information helps users decide whether they should continue or cancel an ongoing rebalance, and keep client teams informed on the expected duration of the rebalance.
+
+Further, having this information readily available and easily accessible via Kubernetes primitives, allows users and third-party tools like the Kubernetes CLI or 3rd party tools to easily track the progression of a partition rebalance.
+
+## Proposal
+Currently, the `KafkaRebalance` custom resource references a `ConfigMap` in the `status.optimizationResult.afterBeforeLoadConfigMap` field which stores the estimated load on the brokers before and after the optimization proposal is executed.
+This proposal extends the status section of the `KafkaRebalance` custom resource to include a `progress` section with a nested `rebalanceProgressConfigMap` field to reference that same `ConfigMap`.
+This `ConfigMap` will also be enhanced to contain information related to an ongoing partition rebalance.
+
+The `progress` section of the `KafkaRebalance` resource will look like the following:
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalance
+metadata:
+  name: my-rebalance
+spec: {}
+status:
+  conditions:
+  - lastTransitionTime: "2024-11-05T15:28:23.995129903Z"
+    status: "True"
+    type: ProposalReady | Rebalancing | Stopped | Ready [1]
+  observedGeneration: 1
+  optimizationResult:
+    afterBeforeLoadConfigMap: my-rebalance
+    dataToMoveMB: 0
+    excludedBrokersForLeadership: []
+    excludedBrokersForReplicaMove: []
+    excludedTopics: []
+    intraBrokerDataToMoveMB: 0
+    monitoredPartitionsPercentage: 100
+    numIntraBrokerReplicaMovements: 0
+    numLeaderMovements: 16
+    numReplicaMovements: 0
+    onDemandBalancednessScoreAfter: 95.4347095948149
+    onDemandBalancednessScoreBefore: 89.4347095948149
+    provisionRecommendation: ""
+    provisionStatus: RIGHT_SIZED
+    recentWindows: 1
+  progress: [1]
+    rebalanceProgressConfigMap: my-rebalance [2]
+```
+[1] The `progress` section will be visible during the `ProposalReady`, `Rebalancing`, `Stopped`, `NotReady` and `Ready` states.
```

Comment:
based on the decision we'll make, I guess we should remove `ProposalReady` from here and in any other place where it's mentioned.

Reply:
Even if we decide to not provide a forecast value for the `estimatedTimeToCompletionInMinutes` field in the `ProposalReady` state, would there be any value still supporting this state for showing `completedDataMovementPercentage` = 0 to emphasize that the rebalance hasn't started? Of course it this state, `estimatedTimeToCompletionInMinutes` would be assigned `-1` or whatever value we decide to represent that it is not available.  What do you think?